kong/tools/uri.lua

```diff
@@ -35,86 +34,133 @@ local RESERVED_CHARACTERS = {
   [0x5D] = true, -- ]
 }
 local TMP_OUTPUT = require("table.new")(16, 0)
+local SLASH = "/"
 local DOT = string_byte(".")
-local SLASH = string_byte("/")
 
 
-function _M.normalize(uri, merge_slashes)
-  table_clear(TMP_OUTPUT)
-
-  -- Decoding percent-encoded triplets of unreserved characters
-  uri = ngx_re_gsub(uri, "%([\\dA-F]{2})", function(m)
+local function percent_decode(m)
     local hex = m[1]
     local num = tonumber(hex, 16)
     if RESERVED_CHARACTERS[num] then
       return string_upper(m[0])
     end
 
     return string_char(num)
-  end, "joi")
+end
 
-  local output_n = 0
 
-  while #uri > 0 do
-    local FIRST = string_byte(uri, 1)
-    local SECOND = string_byte(uri, 2)
-    local THIRD = string_byte(uri, 3)
-    local FOURTH = string_byte(uri, 4)
+local function escape(m)
+  return string_format("%%%02X", string_byte(m[0]))
+end
 
-    if uri == "/." then -- /.
-      uri = "/"
 
-    elseif uri == "/.." then -- /..
-      uri = "/"
-      if output_n > 0 then
-        output_n = output_n - 1
-      end
+function _M.normalize(uri, merge_slashes)
+  if uri == "" or uri == SLASH then
+    return uri
+  end
+
+  -- Decoding percent-encoded triplets of unreserved characters
+  if string_find(uri, "%", 1, true) then
+    uri = ngx_re_gsub(uri, "%([\\dA-F]{2})", percent_decode, "joi")
+  end
 
-    elseif uri == "." or uri == ".." then
-      uri = ""
+  if uri == "/." or uri == "/./" or uri == "/.." or uri == "/../" then
+    return SLASH
+  end
 
-    elseif FIRST == DOT and SECOND == DOT and THIRD == SLASH then -- ../
-      uri = string_sub(uri, 4)
+  if string_find(uri, "/.",  -2, true) == nil and
+     string_find(uri, "/..", -3, true) == nil and
+     string_find(uri, "/./",  1, true) == nil and
+     string_find(uri, "/../", 1, true) == nil
```

Comment:
Two questions about this:
a) Can't the result of these calls to `string_find` be used/reused on the code below? Aren't the subsequent calls to `string_find` redoing work?
b) What is the logic behind doing these checks in this particular order vs the reverse (doing the "start checks" first?

Reply:
I changed this to just check the `.`.